### ISSUE TITLE
leaving the "Find What" and "Replace With" field blank results in error

### ISSUE THREAD

**Issue by spereverde** [collective/collective.searchandreplace#43]
### description

* When leaving **Find What** empty results in an error. Interestingly, the field "Find What" is a required field but the required property doesn't seem to work all the time (the error didn't always happen) . For full traceback, see below.
* When leaving the **Replace With** empty and pressing replace,  also results in an error. For full traceback, see below.
* This seems to happen on all three button pushes (preview, replace and reset when the find what field is empty)
* you can quickly reproduce the error inside the jupyter notebook binder for more traceback and info.

#### traceback empty replace with:

```bash
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module zope.formlib.form, line 868, in __call__
  Module five.formlib.formbase, line 50, in update
  Module zope.formlib.form, line 849, in update
  Module zope.formlib.form, line 640, in success
  Module collective.searchandreplace.browser.searchreplaceform, line 113, in action_replace
  Module collective.searchandreplace.browser.searchreplaceform, line 131, in replace_filtered
  Module collective.searchandreplace.searchreplaceutility, line 151, in replaceFilteredOccurences
  Module collective.searchandreplace.searchreplaceutility, line 246, in replace_occurences_in_object
  Module collective.searchandreplace.searchreplaceutility, line 279, in replaceText
TypeError: coercing to Unicode: need string or buffer, NoneType found
```

#### traceback empty find what:

```bash
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module zope.formlib.form, line 873, in __call__
  Module zope.formlib.form, line 994, in render
  Module zope.formlib.form, line 863, in render
  Module Products.Five.browser.pagetemplatefile, line 125, in __call__
  Module Products.Five.browser.pagetemplatefile, line 59, in __call__
  Module zope.pagetemplate.pagetemplate, line 132, in pt_render
  Module zope.pagetemplate.pagetemplate, line 240, in __call__
  Module zope.tal.talinterpreter, line 271, in __call__
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 888, in do_useMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 858, in do_defineMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 946, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 858, in do_defineMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 954, in do_defineSlot
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 533, in do_optTag_tal
  Module zope.tal.talinterpreter, line 518, in do_optTag
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 858, in do_defineMacro
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 531, in do_optTag_tal
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 531, in do_optTag_tal
  Module zope.tal.talinterpreter, line 513, in no_tag
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 742, in do_insertStructure_tal
  Module Products.PageTemplates.Expressions, line 218, in evaluateStructure
  Module zope.tales.tales, line 696, in evaluate
   - URL: /opt/plone/x0135809/buildout/eggs/collective.searchandreplace-8.0.0-py2.7.egg/collective/searchandreplace/browser/pageform.pt
   - Line 92, Column 16
   - Expression: <PathExpr standard:u'context/@@searchreplacetable'>
   - Names:
      {'args': (),
       'container': <PloneSite at /mptest>,
       'context': <PloneSite at /mptest>,
       'default': <object object at 0x7f830cac9560>,
       'here': <PloneSite at /mptest>,
       'loop': {},
       'nothing': None,
       'options': {},
       'repeat': <Products.PageTemplates.Expressions.SafeMapping object at 0x7f82ec2149f0>,
       'request': <HTTPRequest, URL=http://icts-t-plone-7:11080/mptest/@@searchreplaceform>,
       'root': <Application at >,
       'template': <Products.Five.browser.pagetemplatefile.ViewPageTemplateFile object at 0x7f82f1efd4d0>,
       'traverse_subpath': [],
       'user': <PropertiedUser 'admin'>,
       'view': <Products.Five.metaclass.SearchReplaceForm object at 0x7f82ec6b9f90>,
       'views': <Products.Five.browser.pagetemplatefile.ViewMapper object at 0x7f82ec6a94d0>}
  Module zope.tales.expressions, line 217, in __call__
  Module Products.PageTemplates.Expressions, line 155, in _eval
  Module Products.PageTemplates.Expressions, line 117, in render
  Module Products.Five.browser.metaconfigure, line 485, in __call__
  Module Products.Five.browser.pagetemplatefile, line 125, in __call__
  Module Products.Five.browser.pagetemplatefile, line 59, in __call__
  Module zope.pagetemplate.pagetemplate, line 132, in pt_render
  Module zope.pagetemplate.pagetemplate, line 240, in __call__
  Module zope.tal.talinterpreter, line 271, in __call__
  Module zope.tal.talinterpreter, line 343, in interpret
  Module zope.tal.talinterpreter, line 583, in do_setLocal_tal
  Module zope.tales.tales, line 696, in evaluate
   - URL: /opt/plone/x0135809/buildout/eggs/collective.searchandreplace-8.0.0-py2.7.egg/collective/searchandreplace/browser/searchreplacetable.pt
   - Line 1, Column 0
   - Expression: <PathExpr standard:u'view/getItems'>
   - Names:
      {'args': (),
       'container': <PloneSite at /mptest>,
       'context': <PloneSite at /mptest>,
       'default': <object object at 0x7f830cac9560>,
       'here': <PloneSite at /mptest>,
       'loop': {},
       'nothing': None,
       'options': {},
       'repeat': <Products.PageTemplates.Expressions.SafeMapping object at 0x7f82ecaf42b8>,
       'request': <HTTPRequest, URL=http://icts-t-plone-7:11080/mptest/@@searchreplaceform>,
       'root': <Application at >,
       'template': <Products.Five.browser.pagetemplatefile.ViewPageTemplateFile object at 0x7f82f1f30bd0>,
       'traverse_subpath': [],
       'user': <PropertiedUser 'admin'>,
       'view': <Products.Five.metaclass.SimpleViewClass from /opt/plone/x0135809/buildout/eggs/collective.searchandreplace-8.0.0-py2.7.egg/collective/searchandreplace/browser/searchreplacetable.pt object at 0x7f82ec6982d0>,
       'views': <Products.Five.browser.pagetemplatefile.ViewMapper object at 0x7f82ec698c90>}
  Module zope.tales.expressions, line 217, in __call__
  Module Products.PageTemplates.Expressions, line 155, in _eval
  Module Products.PageTemplates.Expressions, line 117, in render
  Module collective.searchandreplace.browser.searchreplacetable, line 35, in getItems
  Module collective.searchandreplace.searchreplaceutility, line 173, in findObjects
  Module kuleuven.patches.language, line 31, in searchResults
  Module Products.CMFPlone.CatalogTool, line 394, in searchResults
  Module Products.ZCatalog.ZCatalog, line 610, in searchResults
  Module Products.ZCatalog.Catalog, line 925, in searchResults
  Module Products.ZCatalog.Catalog, line 547, in search
  Module Products.ZCTextIndex.ZCTextIndex, line 217, in _apply_index
  Module Products.ZCTextIndex.QueryParser, line 139, in parseQuery
ParseError: Query contains only common words: u'**'
```

**Comment by spereverde** [comment]
I added a check for rtext in this commit: https://github.com/spereverde/collective.searchandreplace/commit/77df79e646f096a6fcaaea650e0b8bb82f36403a
That fixes the replacing with an empty 'replace with' field, but I'm not sure it is the cleanest way to fix it.
If you think it's acceptable, I can make a PR.

**Comment by spereverde** [comment]
I also found a way to make the form not fail when a user does not fill in the `findWhat` field:
https://github.com/spereverde/collective.searchandreplace/commit/96ac0beb6da89431386adf6854aadd383bcc4a42
It will then go back to the form indicating that the findWhat field is required.
Again not sure if it's the best way to fix it, if acceptable I'll make a PR

**Comment by mauritsvanrees** [comment]
I have not tried it, but both fixes seem okay at first glance. A PR would be welcome.
Adding tests that fail without these fixes would be a nice bonus, but I realise tests can be tricky.

**Comment by spereverde** [comment]
Thanks for having a look! I discussed also with @gotcha, one test was already added in my fork, I will try to add the other one as well.
I will make a couple of PR’s for the fixes, you will see them coming next week hopefully.

Met vriendelijke groet,

Kim Paulissen
KU Leuven - ICTS - Faciliteiten voor Inter- en Intranet
W. De Croylaan 52C
3001 Heverlee
België

**Comment by mauritsvanrees** [comment]
I have merged all your pull requests. Thanks for the improvements!
I can do a release. But maybe it is good if you check that all is well now?

**Comment by spereverde** [comment]
Thank you for all the fixes, approvals and merging :)
I will test it on our setup again to be sure and let you know asap
A release would be great then!

**Comment by spereverde** [comment]
Just did several tests in different setups and all seems to be working fine!
So a release would be greatly appreciated :-)
Thanks again!

**Comment by mauritsvanrees** [comment]
I have released [8.1.1](https://pypi.org/project/collective.searchandreplace/8.1.1/). Thanks!
